keps/sig-node/2621-cpu-allocation-llc-affinity/README.md

```diff
@@ -0,0 +1,127 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Caches are not considered in current Kubernetes cpu-manager, in some architectures,  each socket/package owns more than one L3 cache, containers may encounter performance degradation for L3 cache interference and lower hit rate.
+Add support for L3 cache affinity during container cpu allocation, while in the same package/socket, try to use cpus sharing L3 cache  for container demand but not just choose from all cpus in the package/socket.
+
+## Motivation
+
+Kubernetes cpu-manager tries to allocate cpus in the same core, socket/package, gaining better performance.  In traditional architecture, L3 cache is shared between the whole socket, current cpus allocator works well.
+However, the allocation algorithm may encounter problem in processors like 2nd Gen AMD EPYC™,  each ccx(a term used by AMD to describe a cluster of physical cores along with the shared level 3 cache) owns its L3 cache, more than one L3 cache exists in a socket/package. Depending on current cpu allocation may face L3 cache interference. For example, 4 cores with HT in ccx, a container demand for 8 cpus may not get the whole ccx, but get some cpus in other ccx(see figure below), container A and B may affect each other while the other flush l3 cache. In our opinion, container's cpu locality should be considered.
+
+![allocation_motivation](allocation_motivation.png "allocation_motivation")
+
+### Goals
+
+Support L3 cache affinity in cpu allocation in architecture with more than one l3 cache in socket/package.
+
+### Future Work
+
+Cross-die may also decrease process performance. We will add die affinity in the future.
+
+## Proposal
+
+- Add cache id to cadvisor
+In cadvisor PR(https://github.com/google/cadvisor/pull/2847/),  use /sys/devices/system/cpu/cpu*/cache/index3/id to get L3 cache id of current cpu, and store it as cpu topology.
+- Add uncore cache to cadvisor
+In cadvisor PR(https://github.com/google/cadvisor/pull/2849), add L3 cache not shared among the whole socket(uncore cache) to core info in cpu topology.
+
+### User Stories (Optional)
+
+Workload is memory sensitive, this feature can reduce memory(L3 cache) latency.
+Also, we make a bench with stream2 DAXPY, as we can see, cross ccx(cross l3 cache) gets lower bandwidth.
+
+![stream2_daxpy](stream2_daxpy.png "stream2_daxpy")
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+L3 cache affinity will not always get a better performance, however, we do think, workload in containers should not influence other containers. Decreasing L3 cache-miss in individual containers should be taken into consideration during programming workload or use other L3 cache allocation and isolation technology, which are not our topic.
```

Comment:
The possible business case here e.g. performance team found that explicit __builtin_prefetch reduce L3 cache misses in application which uses amount of memory of the same size as L3 cache, but only if nobody else is running on the same core's cluster which shares common L3 cache. And they want to guarantee the same condition on the worker node managed by K8s.
So in this case reasonable to prevision all CPU as exclusive CPU of shared group  to one containers.
E.g. we have 4 cpu in core cluster (they share L3 cache), container requests 3. In this scenario, it implies 4 cpu should be provided, otherwise if only 3 cpu provided to container, 4th cpu could trash the L3 cache, once it provisioning to another container.